miner/scroll_worker.go

```diff
@@ -0,0 +1,943 @@
+			return
+		}
+		header.Coinbase = w.coinbase
+	}
+
+	common.WithTimer(prepareTimer, func() {
+		if err := w.engine.Prepare(w.chain, header); err != nil {
+			log.Error("Failed to prepare header for mining", "err", err)
+			return
+		}
+	})
+
+	// If we are care about TheDAO hard-fork check whether to override the extra-data or not
+	if daoBlock := w.chainConfig.DAOForkBlock; daoBlock != nil {
+		// Check whether the block is among the fork extra-override range
+		limit := new(big.Int).Add(daoBlock, params.DAOForkExtraRange)
+		if header.Number.Cmp(daoBlock) >= 0 && header.Number.Cmp(limit) < 0 {
+			// Depending whether we support or oppose the fork, override differently
+			if w.chainConfig.DAOForkSupport {
+				header.Extra = common.CopyBytes(params.DAOForkBlockExtra)
+			} else if bytes.Equal(header.Extra, params.DAOForkBlockExtra) {
+				header.Extra = []byte{} // If miner opposes, don't let it use the reserved extra-data
+			}
+		}
+	}
+
+	parentState, err := w.chain.StateAt(parent.Root())
+	if err != nil {
+		log.Error("failed to fetch parent state", "err", err)
+		return
+	}
+
+	// fetch l1Txs
+	var l1Messages []types.L1MessageTx
+	if w.chainConfig.Scroll.ShouldIncludeL1Messages() {
+		common.WithTimer(collectL1MsgsTimer, func() {
+			l1Messages = w.collectPendingL1Messages(*rawdb.ReadFirstQueueIndexNotInL2Block(w.eth.ChainDb(), parent.Hash()))
+		})
+	}
+
+	tidyPendingStart := time.Now()
+	// Fill the block with all available pending transactions.
+	pending := w.eth.TxPool().Pending(false)
+	// Split the pending transactions into locals and remotes
+	localTxs, remoteTxs := make(map[common.Address]types.Transactions), pending
+	for _, account := range w.eth.TxPool().Locals() {
+		if txs := remoteTxs[account]; len(txs) > 0 {
+			delete(remoteTxs, account)
+			localTxs[account] = txs
+		}
+	}
+	collectL2Timer.UpdateSince(tidyPendingStart)
+
+	w.currentPipelineStart = time.Now()
+	w.currentPipeline = pipeline.NewPipeline(w.chain, w.chain.GetVMConfig(), parentState, &w.coinbase, header, w.getCCC()).WithBeforeTxHook(w.beforeTxHook)
+	if err := w.currentPipeline.Start(time.Unix(int64(header.Time), 0)); err != nil {
+		log.Error("failed to start pipeline", "err", err)
+		return
+	}
+
+	// Short circuit if there is no available pending transactions.
+	// But if we disable empty precommit already, ignore it. Since
+	// empty block is necessary to keep the liveness of the network.
+	if len(localTxs) == 0 && len(remoteTxs) == 0 && len(l1Messages) == 0 && atomic.LoadUint32(&w.noempty) == 0 {
+		return
+	}
+
+	if w.chainConfig.Scroll.ShouldIncludeL1Messages() && len(l1Messages) > 0 {
+		log.Trace("Processing L1 messages for inclusion", "count", len(l1Messages))
+		txs, err := types.NewL1MessagesByQueueIndex(l1Messages)
+		if err != nil {
+			log.Error("Failed to create L1 message set", "l1Messages", l1Messages, "err", err)
+			return
+		}
+
+		if result := w.currentPipeline.TryPushTxns(txs, w.onTxFailingInPipeline); result != nil {
```

Comment:
Just to make sure I understand: `TryPushTxns` will wait for the trace result but not for the CCC result, right?

Also, by the time `startNewPipeline` returns, it's likely that the block is already finished, right?

Reply:
> TryPushTxns will wait for the trace result but not for the CCC result, right

Correct, but if CCC result happens to come while it is waiting for the trace result, it will return the CCC result to caller.

> Also, by the time startNewPipeline returns, it's likely that the block is already finished, right?

If there are enough txns in the pool to trigger the CCC, certainly. If not, we will wait until the deadline and keep appending the txns that come after `startNewPipeline`returns.